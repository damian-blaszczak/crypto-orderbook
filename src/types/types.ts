export type Feed = {
  ca: string;
  co: number;
  pa: string;
  ra: string;
  sa: string;
};

export type OrderbookProps = {
  bid: Feed[];
  ask: Feed[];
  marketStats: {
    max: number | '-';
    min: number | '-';
  };
  setMarketCode: (value: string) => void;
  marketCode: string;
};

export type MarketStats = {
  h: number;
  l: number;
  m: string;
  r24h: number;
  v: number;
};

export type MarketStatsSnapshotBody = {
  stats: MarketStats;
  status: string;
};

export type OrdersSnapshotBody = {
  buy: Feed[];
  sell: Feed[];
  seqNo: string;
  status: string;
  timestamp: string;
};

export type SnaptshotData = {
  action: string;
  body: OrdersSnapshotBody | MarketStatsSnapshotBody;
  requestId: string;
  statusCode: number;
};

type OrderChanges = {
  action: string;
  entryType: string;
  marketCode: string;
  rate: string;
  state: Feed[];
};

export type PushOrdersData = {
  action: string;
  message: {
    changes: OrderChanges[];
  };
  seqNo: number;
  timestamp: string;
  topic: string;
};

export type PushMarketStatsData = {
  action: string;
  message: MarketStats[];
  seqNo: number;
  timestamp: string;
  topic: string;
};
