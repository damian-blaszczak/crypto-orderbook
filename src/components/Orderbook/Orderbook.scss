.orderbook {
  margin: 16px 24px;
  padding: 16px;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 0px 16px 24px rgba(0, 0, 0, 0.1);
  &__header {
    display: grid;
    grid-template-columns: 33% 33% 33%;
    @media (max-width: 600px) {
      grid-template-columns: 100%;
      grid-template-rows: 50px 50px 50px;
    }
  }
  &__container {
    display: grid;
    grid-template-columns: 50% 50%;
    margin-top: 16px;
    @media (max-width: 800px) {
      grid-template-columns: 100%;
      grid-template-rows: 800px 800px;
    }
  }
  &__header-item {
    display: flex;
    align-items: center;
    &:nth-child(2) {
      justify-content: center;
    }
    &:last-child {
      justify-content: flex-end;
    }
    @media (max-width: 800px) {
      justify-content: center;
      &:last-child {
        justify-content: center;
      }
    }
  }
  &__header-item-title {
    color: #90a4ae;
    margin-right: 4px;
    font-weight: 500;
  }
  &__header-item-value {
    color: #455a64;
    font-weight: 600;
  }
  &__market-stats {
    display: flex;
    flex-direction: column;
    gap: 4px;
    color: #90a4ae;
    font-weight: 500;
    span:last-child {
      color: #455a64;
      font-weight: 600;
      font-size: 1.2rem;
      text-align: right;
    }
  }
  &__market-stats-item {
    display: flex;
    justify-content: space-between;
    gap: 4px;
  }
  &__currency-select {
    border: 0;
    cursor: pointer;
  }
  &__button {
    border-radius: 4px;
    border: 0;
    padding: 8px 16px;
    background-color: lightblue;
    cursor: pointer;
    transition: background-color 250ms;
    &:hover {
      background-color: lightsteelblue;
    }
  }
}
