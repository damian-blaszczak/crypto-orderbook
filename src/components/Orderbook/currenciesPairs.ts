const currenciesPairs = [
  'BTC-PLN',
  'ETH-PLN',
  'AAVE-PLN',
  'ADA-PLN',
  'ALG-PLN',
  'AMLT-PLN',
  'ATRI-PLN',
  'AVAX-PLN',
  'BAT-PLN',
  'BCC-PLN',
  'BCP-PLN',
  'BOB-PLN',
  'BSV-PLN',
  'BTG-PLN',
  'CHZ-PLN',
  'COMP-PLN',
  'DAI-PLN',
  'DASH-PLN',
  'DOGE-PLN',
  'DOT-PLN',
  'ENJ-PLN',
  'EOS-PLN',
  'FTM-PLN',
  'GAME-PLN',
  'GNT-PLN',
  'GRT-PLN',
  'LINK-PLN',
  'LSK-PLN',
  'LML-PLN',
  'LTC-PLN',
  'LUNA-PLN',
  'MANA-PLN',
  'MATIC-PLN',
  'MKR-PLN',
  'NEU-PLN',
  'OMG-PLN',
  'PAY-PLN',
  'REEF-PLN',
  'REP-PLN',
  'SUSHI-PLN',
  'SXP-PLN',
  'TRX-PLN',
  'UNI-PLN',
  'USDC-PLN',
  'USDT-PLN',
  'WIS-PLN',
  'XIN-PLN',
  'XLM-PLN',
  'XRP-PLN',
  'XTZ-PLN',
  'ZEC-PLN',
  'ZRX-PLN'
];

export default currenciesPairs;
