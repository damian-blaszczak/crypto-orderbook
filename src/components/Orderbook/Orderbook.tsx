import OrderbookFeedTable from 'components/OrderbookFeedTable/OrderbookFeedTable';
import { OrderbookProps } from 'types/types';
import './Orderbook.scss';
import { spreadValue } from 'helpers/helpers';
import Loader from 'components/Loader/Loader';
import { useState } from 'react';
import CurrencyDrawer from 'components/CurrencyDrawer/CurrencyDrawer';

const Orderbook = ({ bid, ask, marketStats, setMarketCode, marketCode }: OrderbookProps) => {
  const [isCurrencyDrawerOpen, setCurrencyDrawerOpen] = useState(false);
  return (
    <div className="orderbook">
      <div className="orderbook__header">
        <div className="orderbook__header-item">
          <button
            className="orderbook__button"
            data-testid="currencies-button"
            onClick={() => setCurrencyDrawerOpen(true)}
          >
            {marketCode}
          </button>
        </div>
        <div className="orderbook__header-item">
          <div className="orderbook__header-item-title">Spread:</div>
          <div className="orderbook__header-item-value" data-testid="spreadValue">
            {spreadValue(bid[0]?.ra, ask[0]?.ra)}
          </div>
        </div>
        <div className="orderbook__header-item">
          <div className="orderbook__market-stats">
            <div className="orderbook__market-stats-item">
              <span>Najwyższa (24h)</span> <span>{marketStats.max}</span>
            </div>
            <div className="orderbook__market-stats-item">
              <span>Najniższa (24h)</span> <span>{marketStats.min}</span>
            </div>
          </div>
        </div>
      </div>
      {bid.length && ask.length ? (
        <div className="orderbook__container" data-testid="orderbook-table">
          <OrderbookFeedTable feed={bid} feedType="bid" marketCode={marketCode} />
          <OrderbookFeedTable feed={ask} feedType="ask" marketCode={marketCode} />
        </div>
      ) : (
        <Loader />
      )}
      <CurrencyDrawer
        isOpen={isCurrencyDrawerOpen}
        handleClose={() => setCurrencyDrawerOpen(false)}
        setMarketCode={setMarketCode}
        activeMarketCode={marketCode}
      />
    </div>
  );
};

export default Orderbook;
