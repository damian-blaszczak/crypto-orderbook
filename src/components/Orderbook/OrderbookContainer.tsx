import { useEffect, useRef, useState } from 'react';
import { initWs, subscribeToMarket, unsubscribeToMarket } from 'helpers/helpers';
import Orderbook from 'components/Orderbook/Orderbook';
import {
  Feed,
  PushOrdersData,
  PushMarketStatsData,
  SnaptshotData,
  OrdersSnapshotBody,
  MarketStatsSnapshotBody
} from 'types/types';

const OrderbookContainer = () => {
  const [bid, setBid] = useState<Feed[]>([]);
  const [ask, setAsk] = useState<Feed[]>([]);
  const [updatedBid, setUpdatedBid] = useState<Feed[]>([]);
  const [updatedAsk, setUpdatedAsk] = useState<Feed[]>([]);
  const [marketCode, setMarketCode] = useState('BTC-PLN');
  const [marketStats, setMarketStats] = useState<{ max: number | '-'; min: number | '-' }>({ max: '-', min: '-' });
  const wsRef = useRef<WebSocket | null>(null);
  const preparedBid = [...new Map(bid.map((item) => [item['ra'], item])).values()]
    .slice(0, 25)
    .sort((a, b) => Number(b.ra) - Number(a.ra));
  const preparedAsk = [...new Map(ask.map((item) => [item['ra'], item])).values()]
    .slice(0, 25)
    .sort((a, b) => Number(a.ra) - Number(b.ra));

  const disconnect = () => {
    if (wsRef.current) {
      wsRef.current.close();
      wsRef.current = null;
    }
  };

  const handleOrdersSnapshot = (data: SnaptshotData) => {
    setBid((data.body as OrdersSnapshotBody).buy.sort((a, b) => Number(b.ra) - Number(a.ra)).slice(0, 25));
    setAsk((data.body as OrdersSnapshotBody).sell.sort((a, b) => Number(a.ra) - Number(b.ra)).slice(0, 25));
  };

  const handleMarketStatsSnapshot = (data: SnaptshotData) => {
    const max24h = (data.body as MarketStatsSnapshotBody).stats.h;
    const min24h = (data.body as MarketStatsSnapshotBody).stats.l;
    setMarketStats({ max: max24h, min: min24h });
  };

  const handlePushOrders = (data: PushOrdersData) => {
    setUpdatedBid(
      data.message.changes
        .filter((item) => item.state && item.entryType === 'Buy')
        .map((item) => item.state)
        .flat()
    );
    setUpdatedAsk(
      data.message.changes
        .filter((item) => item.state && item.entryType === 'Sell')
        .map((item) => item.state)
        .flat()
    );
  };

  const handlePushMarketStats = (data: PushMarketStatsData) => {
    const max24h = data.message[0].h;
    const min24h = data.message[0].l;
    setMarketStats({ max: max24h, min: min24h });
  };

  const handleSubscribe = () => {
    setUpdatedBid([]);
    setUpdatedAsk([]);
  };

  const initFeed = (newMarketCode: string) => {
    wsRef.current = initWs(
      newMarketCode,
      disconnect,
      handlePushOrders,
      handlePushMarketStats,
      handleOrdersSnapshot,
      handleMarketStatsSnapshot,
      handleSubscribe
    );
  };

  const connect = () => {
    if (!navigator.onLine) {
      return;
    }
    initFeed(marketCode);
  };

  useEffect(
    () => () => {
      if (wsRef.current) {
        unsubscribeToMarket(wsRef.current, marketCode);
      }
    },
    [marketCode]
  );

  useEffect(() => {
    if (wsRef.current && wsRef.current?.readyState === wsRef.current?.OPEN) {
      subscribeToMarket(wsRef.current, marketCode);
    }
  }, [marketCode]);

  useEffect(() => {
    connect();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (bid.length > 200) {
      const newBid = [...new Map(bid.map((item) => [item['ra'], item])).values()].slice(0, 25);
      setBid(newBid);
    }
    if (ask.length > 200) {
      const newAsk = [...new Map(ask.map((item) => [item['ra'], item])).values()].slice(0, 25);
      setAsk(newAsk);
    }
  }, [bid.length, ask.length, bid, ask]);

  useEffect(() => {
    bid.unshift(...updatedBid);
    ask.unshift(...updatedAsk);
  }, [updatedBid, updatedAsk, bid, ask]);

  return (
    <Orderbook
      bid={preparedBid}
      ask={preparedAsk}
      marketStats={marketStats}
      setMarketCode={setMarketCode}
      marketCode={marketCode}
    />
  );
};

export default OrderbookContainer;
