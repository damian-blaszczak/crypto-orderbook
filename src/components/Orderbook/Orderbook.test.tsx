import { fireEvent, render, screen } from '@testing-library/react';
import { OrderbookProps } from 'types/types';
import Orderbook from './Orderbook';

const props: OrderbookProps = {
  bid: [
    { ra: '195201.1', ca: '0.02', sa: '0.02', pa: '0.02', co: 1 },
    { ra: '195250', ca: '0.01', sa: '0.01', pa: '0.01', co: 1 },
    { ra: '195294.81', ca: '0.00102409', sa: '0.00102409', pa: '0.00102409', co: 1 },
    { ra: '195333', ca: '0.00113728', sa: '0.00113728', pa: '0.00113728', co: 2 },
    { ra: '195400', ca: '0.02535312', sa: '0.02535312', pa: '0.02535312', co: 2 }
  ],
  ask: [
    { ra: '199091.12', ca: '0.15490954', sa: '0.15490954', pa: '0.15490954', co: 1 },
    { ra: '199091.13', ca: '1.529897', sa: '1.529897', pa: '1.529897', co: 1 },
    { ra: '199091.15', ca: '1.0338000', sa: '1.0338000', pa: '1.0338000', co: 1 },
    { ra: '199091.16', ca: '0.0521775', sa: '0.0521775', pa: '0.0521775', co: 1 },
    { ra: '199982.9', ca: '0.2000', sa: '0.2000', pa: '0.2000', co: 1 }
  ],
  marketStats: {
    max: 208499.99,
    min: 197570.01
  },
  setMarketCode: jest.fn(),
  marketCode: 'BTC-PLN'
};

describe('Orderbook', () => {
  it('should render correctly', () => {
    const { container } = render(<Orderbook {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should render spread correctly', () => {
    render(<Orderbook {...props} />);

    const el = screen.getByTestId('spreadValue');
    const spreadValue = (Number(props.ask[0].ra) - Number(props.bid[0].ra)).toFixed(2);

    expect(el).toBeInTheDocument();
    expect(el).toHaveTextContent(spreadValue);
  });

  it('should render orderbook table when there are any bids and asks', () => {
    render(<Orderbook {...props} />);

    expect(screen.getByTestId('orderbook-table')).toBeInTheDocument();
  });

  it('should not render orderbook table when there are not any bids and asks', () => {
    const propsWithoutFeeds = {
      ...props,
      bid: [],
      ask: []
    };
    render(<Orderbook {...propsWithoutFeeds} />);

    expect(screen.queryByTestId('orderbook-table')).not.toBeInTheDocument();
    expect(screen.getByTestId('orderbook-loader')).toBeInTheDocument();
  });

  it('should open currencies drawer on click button and close when currency chosen', async () => {
    render(<Orderbook {...props} />);

    fireEvent.click(screen.getByTestId('currencies-button'));

    expect(screen.getByTestId('currency-drawer').className).toEqual('currencyDrawer currencyDrawer--open');

    fireEvent.click(screen.getAllByTestId('currency-item')[3]);

    expect(screen.getByTestId('currency-drawer').className).toEqual('currencyDrawer ');
  });
});
