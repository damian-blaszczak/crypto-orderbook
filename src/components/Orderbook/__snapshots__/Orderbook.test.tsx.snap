// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orderbook should render correctly 1`] = `
<div>
  <div
    class="orderbook"
  >
    <div
      class="orderbook__header"
    >
      <div
        class="orderbook__header-item"
      >
        <button
          class="orderbook__button"
          data-testid="currencies-button"
        >
          BTC-PLN
        </button>
      </div>
      <div
        class="orderbook__header-item"
      >
        <div
          class="orderbook__header-item-title"
        >
          Spread:
        </div>
        <div
          class="orderbook__header-item-value"
          data-testid="spreadValue"
        >
          3890.02
        </div>
      </div>
      <div
        class="orderbook__header-item"
      >
        <div
          class="orderbook__market-stats"
        >
          <div
            class="orderbook__market-stats-item"
          >
            <span>
              Najwyższa (24h)
            </span>
             
            <span>
              208499.99
            </span>
          </div>
          <div
            class="orderbook__market-stats-item"
          >
            <span>
              Najniższa (24h)
            </span>
             
            <span>
              197570.01
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="orderbook__container"
      data-testid="orderbook-table"
    >
      <div>
        <div
          class="orderbook-header orderbook-header--bid"
        >
          <div
            class="orderbook-header__item"
          >
            <span
              class="orderbook-header__prefix"
            >
              bid
            </span>
             - 
            kupno
          </div>
          <div
            class="orderbook-header__item"
          >
            ilość (
            BTC
            )
          </div>
          <div
            class="orderbook-header__item"
          >
            wartość (
            PLN
            )
          </div>
          <div
            class="orderbook-header__item"
          >
            liczba ofert
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--bid"
        >
          <div
            class="orderbook-row__item"
          >
            195201.1
          </div>
          <div
            class="orderbook-row__item"
          >
            0.02
          </div>
          <div
            class="orderbook-row__item"
          >
            3904.02
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--bid"
        >
          <div
            class="orderbook-row__item"
          >
            195250
          </div>
          <div
            class="orderbook-row__item"
          >
            0.01
          </div>
          <div
            class="orderbook-row__item"
          >
            1952.50
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--bid"
        >
          <div
            class="orderbook-row__item"
          >
            195294.81
          </div>
          <div
            class="orderbook-row__item"
          >
            0.00102409
          </div>
          <div
            class="orderbook-row__item"
          >
            200.00
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--bid"
        >
          <div
            class="orderbook-row__item"
          >
            195333
          </div>
          <div
            class="orderbook-row__item"
          >
            0.00113728
          </div>
          <div
            class="orderbook-row__item"
          >
            222.15
          </div>
          <div
            class="orderbook-row__item"
          >
            2
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--bid"
        >
          <div
            class="orderbook-row__item"
          >
            195400
          </div>
          <div
            class="orderbook-row__item"
          >
            0.02535312
          </div>
          <div
            class="orderbook-row__item"
          >
            4954.00
          </div>
          <div
            class="orderbook-row__item"
          >
            2
          </div>
        </div>
      </div>
      <div>
        <div
          class="orderbook-header orderbook-header--ask"
        >
          <div
            class="orderbook-header__item"
          >
            <span
              class="orderbook-header__prefix"
            >
              ask
            </span>
             - 
            sprzedaż
          </div>
          <div
            class="orderbook-header__item"
          >
            ilość (
            BTC
            )
          </div>
          <div
            class="orderbook-header__item"
          >
            wartość (
            PLN
            )
          </div>
          <div
            class="orderbook-header__item"
          >
            liczba ofert
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--ask"
        >
          <div
            class="orderbook-row__item"
          >
            199091.12
          </div>
          <div
            class="orderbook-row__item"
          >
            0.15490954
          </div>
          <div
            class="orderbook-row__item"
          >
            30841.11
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--ask"
        >
          <div
            class="orderbook-row__item"
          >
            199091.13
          </div>
          <div
            class="orderbook-row__item"
          >
            1.529897
          </div>
          <div
            class="orderbook-row__item"
          >
            304588.92
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--ask"
        >
          <div
            class="orderbook-row__item"
          >
            199091.15
          </div>
          <div
            class="orderbook-row__item"
          >
            1.0338000
          </div>
          <div
            class="orderbook-row__item"
          >
            205820.43
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--ask"
        >
          <div
            class="orderbook-row__item"
          >
            199091.16
          </div>
          <div
            class="orderbook-row__item"
          >
            0.0521775
          </div>
          <div
            class="orderbook-row__item"
          >
            10388.08
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
        <div
          class="orderbook-row orderbook-row--ask"
        >
          <div
            class="orderbook-row__item"
          >
            199982.9
          </div>
          <div
            class="orderbook-row__item"
          >
            0.2000
          </div>
          <div
            class="orderbook-row__item"
          >
            39996.58
          </div>
          <div
            class="orderbook-row__item"
          >
            1
          </div>
        </div>
      </div>
    </div>
    <div
      class="currencyDrawer "
      data-testid="currency-drawer"
    >
      <div
        class="currencyDrawer__closeIcon"
      />
      <div
        class="currencyDrawer__item currencyDrawer__item--selected"
        data-testid="currency-item"
      >
        BTC-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ETH-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        AAVE-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ADA-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ALG-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        AMLT-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ATRI-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        AVAX-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        BAT-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        BCC-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        BCP-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        BOB-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        BSV-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        BTG-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        CHZ-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        COMP-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        DAI-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        DASH-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        DOGE-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        DOT-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ENJ-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        EOS-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        FTM-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        GAME-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        GNT-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        GRT-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        LINK-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        LSK-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        LML-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        LTC-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        LUNA-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        MANA-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        MATIC-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        MKR-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        NEU-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        OMG-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        PAY-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        REEF-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        REP-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        SUSHI-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        SXP-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        TRX-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        UNI-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        USDC-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        USDT-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        WIS-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        XIN-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        XLM-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        XRP-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        XTZ-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ZEC-PLN
      </div>
      <div
        class="currencyDrawer__item "
        data-testid="currency-item"
      >
        ZRX-PLN
      </div>
    </div>
    <div
      class="currencyDrawer__overlay "
    />
  </div>
</div>
`;
